README.md
# Stopwatch

Stopwatch is a simple stopwatch-like timer that can emit events. It currently works only in seconds and will parse a string for seconds, minutes, and hours, and fractions thereof.

## Example usage

### Creating instances
```javascript
// Node
var Stopwatch = require('stopwatch-emitter').Stopwatch;
var stopwatch1 = new Stopwatch('5m');
var stopwatch2 = new Stopwatch('30s');

// Browser
var Stopwatch = new Stopwatch('0.5h');
```

### Main methods
```javascript
var stopwatch = new Stopwatch('5m');

// Start
stopwatch.start();

// Stop
stopwatch.stop();

// Pause
stopwatch.pause();

// Restart from any spot
stopwatch.restart();
```

### Events
Stopwatch implements a classic event emitter. The node version uses the node event emitter and the browser version uses Oliver Caldwell's implementation of [EventEmitter](https://github.com/Wolfy87/EventEmitter) (thanks!). All events have the same name as the method that invokes them. An additional event `tick` is also available which, as you might imagine, is called every second the stopwatch ticks.
```javascript
var stopwatch = new Stopwatch('5m');

stopwatch.on('start', function(){
  console.log('Started!');
});

stopwatch.on('pause', function(){
  console.log('Paused!');
});

stopwatch.start(); // Started!
stopwatch.pause(); // Paused!

stopwatch.on('tick', function(){
  console.log('Tick!');
});

stopwatch.restart();
// Tick!
// Tick!

```

### Other methods
Stopwatch also have a few getters for time remaining, current time, and the max time, all in seconds. It also has an isRunning() call.
```javascript
var stopwatch = new Stopwatch('60s');

stopwatch.getCurrentTime(); // 0
stopwatch.getMaxTime(); // 60

// Start and wait 20 seconds...
stopwatch.start()

stopwatch.getRemainingTime(); // 40
stopwatch.isRunning(); // true
```


CHANGELOG.md
v0.0.4
Date: 05-07-2013
Changes:

  - Fixed a number of edge cases and added unit tests for them
  - Added examples

v0.0.3
Date: 05-06-2013
Changes:

  - Add EventEmitter for browser as submodule

v0.0.2

Date: 05-05-2013
Changes:

  - Added unit tests
  - Fixed prototypal inheritance of EventEmitter
  - Setup npm stuff

v0.0.1

Date: 05-04-2013
Changes:

  - Initial commits

test/karma.conf.js
/* jshint undef: false */


// base path, that will be used to resolve files and exclude;
basePath = '.';
files = [
  JASMINE
  , JASMINE_ADAPTER
  , '../dist/browser/Stopwatch.js'
  , './**/*.spec.js'
];

test/spec/timer.spec.js
/* jshint undef: false */

describe('Stopwatch defaults', function(){
  var stopwatch;

  beforeEach(function(){
    stopwatch = new Stopwatch();
  });

  it('should be an instance of Stopwatch', function(){
    var isInstance = stopwatch instanceof Stopwatch;
    expect(isInstance).toEqual(true);
  });

  it('should start at 0', function(){
    expect(stopwatch.getCurrentTime()).toEqual(0);
  });

  it('should have a default time of 5m (300 seconds)', function() {
    expect(stopwatch.getMaxTime()).toEqual(300);
  });

  it('should fire a start event when started', function(){
    var startFired = false;
    stopwatch.on('start', function(){ startFired = true; });
    stopwatch.start();

    expect(startFired).toBe(true);
  });

  it('should fire a stop event when stopped', function(){
    var stopFired = false;
    stopwatch.on('stop', function(){ stopFired = true; });
    stopwatch.start();
    stopwatch.stop();

    expect(stopFired).toBe(true);
  });

  it('should fire a pause event', function(){
    var pauseFired = false;
    stopwatch.on('pause', function(){ pauseFired = true; });
    stopwatch.start();
    stopwatch.pause();

    expect(pauseFired).toBe(true);
  });

  it('should fire a restart event but not a start event', function(){
    var restartFired = false
    , startFired = false;

    stopwatch.on('restart', function(){ restartFired = true; });
    stopwatch.on('start', function(){ startFired = true; });

    stopwatch.restart();

    expect(restartFired).toBe(true);
    expect(startFired).toBe(false);
  });
});

describe('Stopwatch parseTime', function(){
  // Use a single instance and just place with parseTime
  var stopwatch = new Stopwatch();

  it('should parse seconds correctly', function(){
    stopwatch.parseTime('5s');
    expect(stopwatch.getMaxTime()).toEqual(5);
  });

  it('should parse minutes correctly', function(){
    stopwatch.parseTime('2m');
    expect(stopwatch.getMaxTime()).toEqual(120);
  });

  it('should parse hours correctly', function(){
    stopwatch.parseTime('1h');
    expect(stopwatch.getMaxTime()).toEqual(3600);
  });

  it('should also parse floats correctly', function(){
    stopwatch.parseTime('0.5h');
    expect(stopwatch.getMaxTime()).toEqual(1800);
  });
});

describe('Stopwatch behavior', function(){
  var stopwatchCallback;

  beforeEach(function(){
    stopwatchCallback = jasmine.createSpy('stopwatchCallback');
    jasmine.Clock.useMock();
  });

  it('should stop at maxTime and fire callback', function(){
    var stopwatch = new Stopwatch('5s');

    stopwatch.on('stop', function(){ stopwatchCallback(); });
    expect(stopwatchCallback).not.toHaveBeenCalled();

    stopwatch.start();
    // Tick 5 seconds
    jasmine.Clock.tick(5000);

    expect(stopwatchCallback).toHaveBeenCalled();
  });

  it('should call tick callback correctly', function(){
    var totalTicks = 0;
    var stopwatch = new Stopwatch('5s');

    stopwatch.on('tick', function(){
      totalTicks++;
    });

    stopwatch.start();

    jasmine.Clock.tick(5000);
    expect(totalTicks).toEqual(5);
  });

  it ('should not emit "stop" unless currently running', function(){
    var stopCalled = 0;
    var stopwatch = new Stopwatch('5s');

    stopwatch.on('stop', function(){
      stopCalled++;
    });

    stopwatch.start();
    jasmine.Clock.tick(1000);
    stopwatch.stop();
    stopwatch.stop();

    expect(stopCalled).toEqual(1);
  });

  it('should only emit "stop" once when method is called and once when timer runs out', function(){
    var stopCalled = 0;
    var stopwatch = new Stopwatch('5s');

    stopwatch.on('stop', function(){
      stopCalled++;
    });

    stopwatch.start();
    jasmine.Clock.tick(4000);
    stopwatch.stop();

    expect(stopCalled).toEqual(1);

    stopwatch.restart();
    jasmine.Clock.tick(5000);

    expect(stopCalled).toEqual(2);
  });
});

describe('Stopwatch getters and booleans', function() {
  var stopwatch;

  beforeEach(function(){
    stopwatch = new Stopwatch('60s');
    jasmine.Clock.useMock();
  });

  it('should report the correct getters for start, max, and remaining', function(){
    stopwatch.start();
    jasmine.Clock.tick(5000);

    expect(stopwatch.getCurrentTime()).toEqual(5);
    expect(stopwatch.getRemainingTime()).toEqual(55);
    expect(stopwatch.getMaxTime()).toEqual(60);
  });

  it('should correctly report if it is running or not', function(){
    expect(stopwatch.isRunning()).toBe(false);

    stopwatch.start();
    expect(stopwatch.isRunning()).toBe(true);
  });

  it('should correctly report if is not running when paused', function(){
    expect(stopwatch.isRunning()).toBe(false);

    stopwatch.start();
    jasmine.Clock.tick(5000);
    stopwatch.pause();

    expect(stopwatch.isRunning()).toBe(false);
  });

  it('should correctly report if it not running when stopped', function(){
    stopwatch.start();
    jasmine.Clock.tick(5000);
    stopwatch.stop();
    expect(stopwatch.isRunning()).toBe(false);

    stopwatch.restart();
    jasmine.Clock.tick(60000);
    expect(stopwatch.isRunning()).toBe(false);
  });

  it('should ignore additional "start" calls', function(){
    stopwatch.start();
    jasmine.Clock.tick(1000);
    stopwatch.start();
    stopwatch.start();
    expect(stopwatch.getCurrentTime()).toEqual(1);
  });

  it('should clear any queued timeouts when "stop" or "pause" is called', function(){
    stopwatch.start();
    jasmine.Clock.tick(1000);
    stopwatch.stop();
    expect(stopwatch.getCurrentTime()).toEqual(0);

    stopwatch.start();
    jasmine.Clock.tick(1000);
    stopwatch.pause();
    expect(stopwatch.getCurrentTime()).toEqual(1);
  });

  it('should clear any timeouts when "restart" is called', function(){
    stopwatch.start();
    jasmine.Clock.tick(1000);
    stopwatch.restart();
    expect(stopwatch.getCurrentTime()).toEqual(0);

    jasmine.Clock.tick(5000);
    expect(stopwatch.getCurrentTime()).toEqual(5);
  });
});

Gruntfile.js
module.exports = function(grunt) {
  grunt.loadNpmTasks('grunt-karma');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json')

    , meta: {
      banner: (function(){
        var node = [
            "/** <%= pkg.name %> - v<%= pkg.version %> - <%= pkg.homepage %>\n"
          , "  * Copyright (c) <%= grunt.template.today('yyyy') %> <%= pkg.author %>. All rights reserved.\n"
          , "  * Licensed <%= _.pluck(pkg.licenses, 'type')[0] %> - <%= _.pluck(pkg.licenses, 'url')[0] %>\n"
        ]
        , browser = node.concat([
            "  *\n"
          , "  * EventEmitter - git.io/ee\n"
          , "  * Oliver Caldwell\n"
          , "  * MIT license\n"
        ]);

        return {
            node: node.concat(["  */\n"]).join('')
          , browser: browser.concat(["  */\n"]).join('')
        };
      }())
    }

    , paths: {
        lib: './lib'
      , dist: './dist'
      , test: './test'
    }

    , concat: {
      app: {
        files: {
            "<%= paths.dist %>/node/Stopwatch.js": [ "<%= paths.lib %>/NodeEmitter.js", "<%= paths.lib %>/Stopwatch.js" ]
          , "<%= paths.dist %>/browser/Stopwatch.js": [ "<%= paths.lib %>/EventEmitter/EventEmitter.js", "<%= paths.lib %>/Stopwatch.js" ]
        }
      }

      , build_node: {
          options: { banner: "<%= meta.banner.node %>" }
        , files: {
          "<%= paths.dist %>/node/Stopwatch.js": [ "<%= paths.dist %>/node/Stopwatch.js" ]
        }
      }

      , build_browser: {
        options: {
            banner: "<%= meta.banner.browser %>"
            // Strip EventEmitter banner since we are manually re-adding it
          , process: function(src) {
            return src.replace(/(^|\n)\/([\*\s])+EventEmitter([\s\w\.\/\*-]+)([@\w\s]+)\*\//g, '');
          }
        }

        , files: {
            "<%= paths.dist %>/browser/Stopwatch.js": [ "<%= paths.dist %>/browser/Stopwatch.js" ]
          , "<%= paths.dist %>/browser/Stopwatch.min.js": [ "<%= paths.dist %>/browser/Stopwatch.min.js" ]
        }
      }
    }

    , watch: {
      js: {
          files: [
            "<%= paths.lib %>/EventEmitter/EventEmitter.js"
            , "<%= paths.lib %>/Stopwatch.js"
          ]
        , tasks: [ "concat:app", "jshint", "uglify", "karma:unit:run" ]
      }

      , test: {
          files: [ "<%= paths.test %>/**/*.spec.js" ]
        , tasks: [ "karma:unit:run" ]
      }
    }

    , uglify: {
      app: {
        options: {
          // Default compress options. Listed for reference.
          compress: {
            loops          : true
            , unused       : true
            , unsafe       : true
            , cascade      : true
            , warnings     : true
            , booleans     : true
            , evaluate     : true
            , dead_code    : true
            , join_vars    : true
            , if_return    : true
            , sequences    : true
            , hoist_vars   : false
            , hoist_funs   : true
            , properties   : true
            , comparisons  : true
            , conditionals : true
          }

          , mangle: {
            except: [ "Stopwatch", "EventEmitter" ]
          }
        }
        , files: {
          "<%= paths.dist %>/browser/Stopwatch.min.js": [ "<%= paths.dist %>/browser/Stopwatch.js" ]
        }

      }
    }

    , jshint: {
        options: { jshintrc: "./.jshintrc" }
      , all: [
        "<%= paths.lib %>/EventEmitter/EventEmitter.js"
        , "<%= paths.lib %>/Stopwatch.js"
      ]
    }

    , karma: {
      unit: {
        configFile: '<%= paths.test %>/karma.conf.js'
        , runnerPort: 9999
        , autoWatch: false
        , browsers: ['Chrome']
      }
    }
  });

  grunt.registerTask('default', [
    "concat:app"
    , "jshint"
    , "uglify"
    , "concat:build_node"
    , "concat:build_browser"
    , "karma:unit:run"
  ]);
};

lib/Timer.js
/* global EventEmitter, Timer */

(function(exports){
  var hasProp = {}.hasOwnProperty
    , _extends = function(child, parent) {
      for (var key in parent) {
        if (hasProp.call(parent, key)) child[key] = parent[key];
      }

      function ctor() {
        this.constructor = child;
      }

      ctor.prototype = parent.prototype;
      child.prototype = new ctor();
      child.__super__ = parent.prototype;
      return child;
    };

  var Timer = (function(_super){
    _extends(Timer, _super);

    function Timer(maxTime) {
      if (maxTime == null) maxTime = '5m';
      this.currentTime = 0;
      this.parseTime(maxTime);
    }

    var proto = Timer.prototype;

    proto.parseTime = function(time) {
      time = time.toString().match(/([\d\.]+)(\w{1})/);

      var timeValue = parseFloat(time[1], 10)
        , timeInterval = time[2];

      switch (timeInterval) {
        case 's':
          timeValue *= 1;
          break;
        case 'm':
          timeValue *= 60;
          break;
        case 'h':
          timeValue *= 60 * 60;
      }
      this.maxTime = Math.round(timeValue);
    };

    proto.tick = function() {
      var self = this;

      if (self.currentTime >= self.maxTime) {
        self.emit('stop');
        return;
      }

      if (!self.running) {
        self.emit('pause');
        return;
      }

      setTimeout(function(){
        self.currentTime++;
        self.tick();
      }, 1000);
    };

    proto.pause = function() {
      this.running = false;
      this.emit('pause');
    };

    proto.stop = function() {
      this.running = false;
      this.currentTime = 0;
      this.emit('stop');
    };

    proto.restart = function() {
      this.currentTime = 0;
      this.start(false);
      this.emit('restart');
    };

    proto.start = function(emit) {
      if (emit == null) emit = true;
      if (emit) this.emit('start');
      this.running = true;
      this.tick();
    };

    proto.getCurrentTime = function() {
      return this.currentTime;
    };

    proto.getRemainingTime = function() {
      return this.maxTime - this.currentTime;
    };

    proto.getMaxTime = function() {
      return this.maxTime;
    };

    return Timer;
  }(EventEmitter));

  exports.Timer = Timer;
}(this));

examples/js/sw.js
/* global angular */

var app = angular.module('Stopwatch', []);

app.run([
  '$rootScope'
  , function($rootScope) {
    $rootScope.safeApply = function(fn) {
      var phase = this.$root.$$phase
        , fn = fn || function() {};

      if (phase === '$apply' || phase === '$digest') {
        fn();
      } else {
        this.$apply(fn);
      }
    };
  }
]);

app.controller('StopwatchCtrl', [
  '$scope', '$timeout'
  , function($scope, $timeout) {
    $scope.timeUnits = [
        {'value': 's', 'name': 'Seconds'}
      , {'value': 'm', 'name': 'Minutes'}
      , {'value': 'h', 'name': 'Hours'}
    ];

    $scope.timeUnit = $scope.timeUnits[0].value;
    $scope.timeValue = 5;
    $scope.stopwatch = createStopwatch();

    $scope.refreshStopwatch = function() {
      $scope.stopwatch = createStopwatch();
    };

    $scope.timeLabel = function() {
      var len = $scope.timeUnits.length
        , i = 0;

      while (i < len) {
        var unit = $scope.timeUnits[i];
        if (unit.value === $scope.timeUnit) return unit.name;
        i++;
      }
    };

    $scope.currentTime = function() {
      return $scope.stopwatch.getCurrentTime();
    };

    $scope.isRunning = function() {
      return $scope.stopwatch.isRunning();
    };

    $scope.start = function() {
      $scope.stopwatch.start();
    };

    $scope.startDisabled = function() {
      return ($scope.stopwatch.getCurrentTime() >= $scope.stopwatch.getMaxTime());
    };

    function emitAddClass(type) {
      var method = type + "Fired";
      $scope[method] = true;

      $timeout(function(){
        $scope[method] = false;
      }, 400);
    }

    function createStopwatch() {
      var time = "" + $scope.timeValue + $scope.timeUnit;
      var stopwatch = new Stopwatch(time);

      stopwatch.on('start', function(){ $scope.safeApply(emitAddClass('start')); });
      stopwatch.on('stop', function(){ $scope.safeApply(emitAddClass('stop')); });
      stopwatch.on('pause', function(){ $scope.safeApply(emitAddClass('pause')); });
      stopwatch.on('restart', function(){ $scope.safeApply(emitAddClass('restart')); });
      stopwatch.on('tick', function(){ $scope.safeApply(); });

      return stopwatch;
    }
  }
]);

lib/Stopwatch.js
/* global EventEmitter, Timer */

(function(exports){
  var hasProp = {}.hasOwnProperty
    , _extends = function(child, parent) {
      for (var key in parent) {
        if (hasProp.call(parent, key)) child[key] = parent[key];
      }

      function ctor() {
        this.constructor = child;
      }

      ctor.prototype = parent.prototype;
      child.prototype = new ctor();
      child.__super__ = parent.prototype;
      return child;
    };

  var Stopwatch = (function(_super){
    _extends(Stopwatch, _super);

    function Stopwatch(maxTime) {
      // Set defaults
      if (maxTime == null) maxTime = '5m';
      this.currentTime = 0;
      this._stopEmitted = false;
      this._running = false;
      this.toQ = [];

      this.parseTime(maxTime);
      this._setupEvents();
    }

    var proto = Stopwatch.prototype;

    // Translate strings into seconds and set maxTime
    proto.parseTime = function(time) {
      // Create an override for numbers. We'll expect seconds.
      if (!isNaN(time)) return this.maxTime = time;

      time = time.toString().match(/([\d\.]+)(\w{1})/);

      var timeValue = parseFloat(time[1], 10)
        , timeInterval = time[2];

      switch (timeInterval) {
        case 's':
          timeValue *= 1;
          break;
        case 'm':
          timeValue *= 60;
          break;
        case 'h':
          timeValue *= 60 * 60;
      }
      this.maxTime = Math.round(timeValue);
    };

    // Setup some instance events
    proto._setupEvents = function() {
      var self = this;

      // Stop emitted tells us we specifically called stop so the we don't
      // emit twice.
      self.on('stop', function(){ self._stopEmitted = true; self.clear(); });

      // Make sure to toggle stopEmitted we any other event fires
      // Also, clear any timeouts if we are already running.
      var stopCb = function() {
        self._stopEmitted = false;
        if (self._running) self.clear();
      };

      self.on('start', stopCb);
      self.on('pause', stopCb);
      self.on('restart', stopCb);
    };

    proto.tick = function() {
      var self = this;

      if (self.currentTime >= self.maxTime) {
        if (!self._stopEmitted) self.emit('stop'); // If we reach the end
        self._running = false; // Make sure we're not running and return
        return;
      }

      // Make sure nothing naughty gets through
      if (!self._running) return;

      // Emit the tick
      self.emit('tick');

      // Store the timeout so we can clear it later if necessary
      var tO = setTimeout(function(){
        self.currentTime++;
        self.tick();
      }, 1000);

      // Push to timeout queue
      self.toQ.push(tO);
    };

    // Clear the timeout queue
    proto.clear = function() {
      var q = this.toQ;

      while(q.length) {
        clearTimeout(q[q.length - 1]);
        q.pop();
      }
    };

    proto.pause = function() {
      if (!this._running) return;
      this._running = false;
      this.emit('pause');
    };

    proto.stop = function() {
      if (!this._running) return;
      this._running = false;
      this.currentTime = 0;
      this.emit('stop');
    };

    proto.restart = function() {
      this.currentTime = 0;
      this.emit('restart');
      this.start(false);
    };

    proto.start = function(emit) {
      if (emit == null) emit = true;
      if ((this._running && emit) || this.currentTime >= this.maxTime) return;
      if (emit) this.emit('start');
      this._running = true;
      this.tick();
    };

    proto.getCurrentTime = function() {
      return this.currentTime;
    };

    proto.getRemainingTime = function() {
      return this.maxTime - this.currentTime;
    };

    proto.getMaxTime = function() {
      return this.maxTime;
    };

    proto.isRunning = function() {
      return !!this._running;
    };

    return Stopwatch;
  }(EventEmitter));

  exports.Stopwatch = Stopwatch;
}(this));